Data/Aeson/TH.hs

```diff
@@ -349,11 +354,35 @@ consToValue target jc opts instTys cons = do
               , let conName = constructorName con
               ]
         | otherwise -> [argsToValue target jc tvMap opts True con | con <- cons]
+    lets :: ToJSONFun -> ExpQ -> ExpQ
+    lets Value e = e
+    lets (Encoding r) e = do
+      e' <- e -- collect used keys
+      labels <- runIO $ readIORef r
+      let decs = map
+            (\k -> let str = BL.unpack $ B.toLazyByteString $ EB.string k
+                  in valD (varP $ varOfKey k) (normalB [| S.pack str |]) [])
+            (Set.toList labels)
+      letE decs $ pure e'
+
+varOfKey :: String -> Name
+varOfKey l = mkName $ "bs_" ++ l
```

Comment:
Also I'm not a fan. This should be using `newName`, as it's local definition, isn't it?

Reply:
Aa problem is that TH-generated definitions also cause shadowed-name etc warnings (as they should). And given that `aeson` is used "everywhere", let's not take any risks.

The `IORef` is another kind of a problem. This `TH` code is already quite difficult to follow. Let's not make it worse. The TH code itself doesn't have performance constraints, only the code we generate with it.

Reply:
Thank you, I've replaced `mkName` with `newName`.